### ISSUE TITLE
extern procs and chpl_external_array

### ISSUE THREAD

**Issue by mppf** [chapel-lang/chapel#16135]
Today, we have `c_ptrTo(myArray)` compiling for rectangular arrays - it returns a pointer to the first element. We also handle `extern proc(x: [] int)` (for example) by interpreting it as `x: c_ptr(int)`.

However, these features predate `chpl_external_array` and `chpl_opaque_array`. Additionally, I am concerned that these features imply that DefaultRectangular arrays must always have the same layout that they do now (and in particular that they must have the same layout on all platforms). 

Besides that, I have in the past had problems because `c_ptrTo(myArray)` does not create a pointer to the array record itself but rather a pointer to the first element. While that might be useful in some cases, I had problems when I was trying to write C code to "move" a value. While we probably will end up with library functions to move a value around, the discrepancy here bothers me, since it makes the `c_ptrTo` function less consistent and less C-like.

Finally, `c_ptrTo(myBlockArray)` will compile today (a Block array is rectangular), but it will almost certainly lead to core dumps if used.

https://github.com/chapel-lang/chapel/blob/aeb77df3db3ca8d224567208ee90d7d0040ac712/modules/internal/CPtr.chpl#L423-L437

https://github.com/chapel-lang/chapel/blob/aeb77df3db3ca8d224567208ee90d7d0040ac712/compiler/passes/expandExternArrayCalls.cpp#L34-L60

Could these features be deprecated and instead rely on features built from `chpl_external_array` and `chpl_opaque_array`?

`chpl_external_array` can handle passing the array bounds and strides in one argument, so that it can handle multi-dimensional arrays. It is plausible that `chpl_external_array` will always be able to represent a DefaultRectangular array (since it could handle for example changing from column-major to row-major or changes like rounding up row sizes to cache lines).

In some cases, it might be useful to pass `chpl_opaque_array` to an `extern` function.

So, should the example declaration
``` chapel
extern proc print_array(x: [] int, n: int);
```

instead be written like this:
``` chapel
extern proc print_array(x: chpl_external_array);
```

or like this:
``` chapel
extern proc print_array(x: chpl_opaque_array);
```

depending on what the C function needs?

Then the compiler could support implicit conversions for passing Chapel arrays to it, so that e.g. `print_array([1,2,3])` will work.

We could additionally provide something like `c_ptrToArrayElements` that always returns a pointer to the first element and only works for 1D arrays that have contiguous layout. For something like LAPACK, we could provide something like `c_ptrToArrayElementsRowMajor` that matches `c_ptrToArrayElements` for 1D arrays and for 2D arrays asserts that the matrix is stored in a particular fashion. I suspect the most reasonable way to handle LAPACK's requirements would be to convert the input arrays to `chpl_external_array` and then compute the appropriate arguments from that (and validate any requirements LAPACK has - such as the rows being tightly packed).

**Comment by bradcray** [comment]
> Finally, c_ptrTo(myBlockArray) will compile today (a Block array is rectangular), but it will almost certainly lead to core dumps if used.

I thought we had planned for `c_ptrTo()` to give a compile-time and/or execution-time error when used on arrays that are not stored contiguously in memory, but in looking at a recent user issue, it seemed that that was not the case (i.e., even if we'd planned it, we never implemented it, or broke it at some point).  In any case, I think we could and should do so regardless of the rest of this discussion.

I remain pretty iffy on the whole `chpl_external_array` / `chpl_opaque_array` capability and story, and feel it's long overdue for more of a design discussion before committing to it.  That said, I agree with the general theme of "Sometimes I just want to pass C a pointer to my array's elements (when that makes sense)" vs. "Sometimes I may want to have C know more about the meta-data of the array and to get the array record or something equivalent to it."  But I think there's a lot of design space in option 2 to wrestle through.

**Comment by mppf** [comment]
Simply removing the `c_ptrTo` special behavior for arrays and instead changing calls to it to use a different function, like `c_ptrToArrayElements`, would make me much happier with the design. If we can agree on that much it might be worth doing before wrestling with related array interop questions.

**Comment by riftEmber** [comment]
Can this be closed as covered by https://github.com/chapel-lang/chapel/issues/18016?

**Comment by lydia-duncan** [comment]
I don't think so - I think the discussion about handling with interoperability is important and not resolved by that question

**Comment by riftEmber** [comment]
Noting that with the addition of `c_addrOf` as an option that avoids `c_ptrTo`'s special behavior on arrays, there isn't any more CTypes work related to this story planned for 2.0; please let me know if that sounds wrong.

**Comment by mppf** [comment]
@riftEmber - yes I think we are in an OK place for 2.0 regarding this issue. We can view the use of extern procs with `chpl_external_array` / `chpl_opaque_array` as unstable and needing future work.

Could you check one thing?

> Finally, c_ptrTo(myBlockArray) will compile today (a Block array is rectangular), but it will almost certainly lead to core dumps if used.

Could you check that `c_ptrTo(myBlockArray)` gives a compilation error for now? Thanks

**Comment by riftEmber** [comment]
@mppf Sounds good. I confirmed that the following errors on main:
```
use BlockDist;
use CTypes;

param n = 100;
const D = {1..n} dmapped Block(boundingBox={1..n});
var A: [D] real;
writeln(A);
var ptr = c_ptrTo(A); // also errors with c_ptrToConst
```

With error message: `error: Only single-locale rectangular arrays support c_ptrTo() at present`

Here's the code responsible for catching this case: https://github.com/chapel-lang/chapel/blob/0d1cb968afc46bd60184513aaf4e5b4f26493212/modules/standard/CTypes.chpl#L765-L766
